kge/model/embedder/lookup_embedder.py

```diff
@@ -61,21 +61,16 @@ def __init__(
     def prepare_job(self, job: Job, **kwargs):
         super().prepare_job(job, **kwargs)
         if self.normalize_p > 0:
-
-            def normalize_embeddings(job):
-                if self.normalize_with_grad:
-                    self._embeddings.weight = torch.nn.functional.normalize(
-                        self._embeddings.weight, p=self.normalize_p, dim=-1
-                    )
-                else:
-                    with torch.no_grad():
-                        self._embeddings.weight = torch.nn.Parameter(
-                            torch.nn.functional.normalize(
-                                self._embeddings.weight, p=self.normalize_p, dim=-1
-                            )
+            def normalize_embeddings(job : "TrainingJob", trace_entry: Dict[str,str]=None):
+                with torch.no_grad():
+                    self._embeddings.weight.data =\
+                        torch.nn.functional.normalize(
+                            self._embeddings.weight.data, p=self.normalize_p, dim=-1
                         )
-
+            # normalize before starting a batch
             job.pre_batch_hooks.append(normalize_embeddings)
+            # normalize after an epoch (before validation and checkpoint is saved)
+            job.post_epoch_hooks.append(normalize_embeddings)
```

Comment:
More suitable would be post_batch_hook. Otherwise looks good.

Reply:
You mean pre_batch_hooks + post_batch_hooks ?

Reply:
@rufex2001 commented 
> I'm guessing just normalizing after the updates, so post batch only?

This would mean that the computation for the first batch after resuming or initializaton would be wrong. Or what do you think?

Reply:
One should normalize anyway during __init__. This handles initialization and loading: the model starts with correct embeddings. Then they are renormalized whenever they are changed, i.e., after each batch.

Reply:
See https://github.com/uma-pi1/kge/pull/124 to enable this change.